spec.html

```diff
@@ -6419,135 +6419,31 @@ <h1>GetGlobalObject ( )</h1>
+    <h1>Jobs and Host Operations to Enqueue Jobs</h1>
+    <p>A <dfn>Job</dfn> is an abstract operation that initiates an ECMAScript computation when no other ECMAScript computation is currently in progress. A Job abstract operation may be defined to accept an arbitrary set of job parameters.</p>
 
-    <emu-clause id="sec-enqueuejob" aoid="EnqueueJob">
-      <h1>EnqueueJob ( _queueName_, _job_, _arguments_ )</h1>
-      <p>The EnqueueJob abstract operation requires three arguments: _queueName_, _job_, and _arguments_. It performs the following steps:</p>
-      <emu-alg>
-        1. Assert: Type(_queueName_) is String and its value is the name of a Job Queue recognized by this implementation.
-        1. Assert: _job_ is the name of a Job.
-        1. Assert: _arguments_ is a List that has the same number of elements as the number of parameters required by _job_.
-        1. Let _callerContext_ be the running execution context.
-        1. Let _callerRealm_ be _callerContext_'s Realm.
-        1. Let _callerScriptOrModule_ be _callerContext_'s ScriptOrModule.
-        1. Let _pending_ be PendingJob { [[Job]]: _job_, [[Arguments]]: _arguments_, [[Realm]]: _callerRealm_, [[ScriptOrModule]]: _callerScriptOrModule_, [[HostDefined]]: *undefined* }.
-        1. Perform any implementation or host environment defined processing of _pending_. This may include modifying the [[HostDefined]] field or any other field of _pending_.
-        1. Add _pending_ at the back of the Job Queue named by _queueName_.
-        1. Return NormalCompletion(~empty~).
-      </emu-alg>
+    <p>Jobs are scheduled for execution by ECMAScript host environments with abstract operations such as HostEnqueuePromiseJob. Such operations accept two parameters: _job_, an abstract operation, and _arguments_, a list of ECMAScript values. These abstract operations schedule _job_ to be performed with _arguments_ at some future time. Their implementations must conform to the following requirements:</p>
+
+    <ul>
+      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must:
+        <ol>
+          <li>Push an execution context onto the execution context stack.</li>
+          <li>Perform any implementation-defined preparation steps.</li>
+          <li>Perform the abstract operation given by _job_, passing the arguments _arguments_.</li>
+          <li>Perform any implementation-defined cleanup steps.</li>
+          <li>Pop the previously-pushed execution context from the execution context stack.</li>
+        </ol>
+      </li>
+      <li>Only one Job may be actively undergoing evaluation at any point in time.</li>
+      <li>Once evaluation of a Job starts, it must run to completion before evaluation of any other Job starts.</li>
+    </ul>
+
+    <emu-clause id="sec-hostenqueuepromisejob" aoid="HostEnqueuePromiseJob">
+      <h1>HostEnqueuePromiseJob ( _job_, _arguments_ )</h1>
```

Comment:
So sometime between [@littledan's presentation](https://docs.google.com/presentation/d/1LLIIwZyQgm6rbAyWbRjFFSNKlA2DPqslpN8sVKjchjE/edit#slide=id.g54c32c9555_0_172) and this PR,
```HostEnqueueJob("Foo", _job_, _args_)```
got replaced with
```HostEnqueueFooJob(_job_, _args_)```
(roughly speaking).

Why's that?

It seems like defining a single job-enqueuing operation would be simpler than defining a 'class' of them.

Reply:
We discussed the two alternatives that the slides presented in the TC39 meeting, and the consensus there (at least shared by me, @domenic, and @erights, if I understood correctly; I don't remember hearing opinions from others) was that hook-per-job is easier to understand and more flexible. This way, we can define different invariants for different hooks. For example, WeakRef finalizer callbacks should not be FIFO but rather considered unordered, whereas Promise-related job hooks should be ordered.

Reply:
> We discussed the two alternatives that the slides presented

(I'm not positive we're talking about the same thing, because I don't see the slides presenting what I'm suggesting. Are you thinking of the "One-off callbacks" alternative? I'm not.)

Anyhow, it seems to me the spec could as easily attach an invariant to a particular argument value as to a particular operation. E.g.: `Jobs enqueued with _kind_ = "Promise" must be scheduled in FIFO order.` No difference in flexibility there.

As for the "easier to understand" claim, I don't get it. I mean, I could see it if different kinds of jobs needed to supply different extra information when they were enqueued. But the PR says that every job-enqueuing operation accepts the same two parameters.

Reply:
Yes, I was talking about PR #735 vs one-off callbacks (this PR). I am not sure what other option you are thinking of.

I agree, we could do that. We discussed these two options in TC39 and settled on one-off callbacks as subjectively simpler. The decision here is editorial, and we can iterate on it over time.

Reply:
> one-off callbacks (this PR)

Well, now I'm really confused, because [one-off callbacks](https://docs.google.com/presentation/d/1LLIIwZyQgm6rbAyWbRjFFSNKlA2DPqslpN8sVKjchjE/edit#slide=id.g5700ec0e5c_0_4) ("No use of Jobs") doesn't appear to be what this PR is doing.

Reply:
OK, you're right, this PR takes a sort of middle path, where we do pass the abstract operation (which we refer to as a "job") up to the host, which is expected to call it, rather than the host knowing which abstract operation to call back. Anyway, I think we're still safely in the realm of small editorial differences. Is there a problem you see with this version, or do you just see opportunities for rephrasing?

Reply:
> I think we're still safely in the realm of small editorial differences.

I agree it's editorial, but it's an operation that will be used by 'downstream' specs, so a change in its signature has more effect than rephrasing some prose.

> Is there a problem you see with this version

I think the `HostEnqueueJob("Foo", _job_, _args_)` of your presentation would be clearer than the `HostEnqueueFooJob(_job_, _args_)` of this PR. So I asked why that change happened. But I still don't know how much (if any) of what you've said upthread is talking about *that* distinction, vs how much is instead talking about the presentation's distinction between `HostEnqueueJob` and one-off callbacks.

Reply:
@jmdyck The motivation I documented in https://github.com/tc39/ecma262/pull/1597#discussion_r296753876 applies to this PR; it is a quality in common between the second alternative in the slide deck and this PR.

Reply:
So here's what I think you're saying: the consensus was that the presentation's second alternative (one-off callbacks, no use of jobs) was more flexible and easier to understand than the first (`HostEnqueueJob`), but rather than adopt one-off callbacks, you moved (for the purposes of this PR) part-way in that direction, from `HostEnqueueJob` to `HostEnqueueFooJob`. Have I got that right?

Reply:
If I understand you correctly, yes.

Reply:
Okay, good. But that still leaves me puzzled on 2 fronts:

1) If one-off callbacks were deemed better, why not just adopt one-off callbacks?

2) I still don't get why it's deemed more flexible and easier to understand.

You gave the example of specifying invariants, and I pointed out that invariants can be attached regardless. Or consider the example of [a one-off callback](https://docs.google.com/presentation/d/1LLIIwZyQgm6rbAyWbRjFFSNKlA2DPqslpN8sVKjchjE/edit#slide=id.g5700ec0e5c_0_4) from the presentation; I can imagine defining the operation:
```
    HostCleanupFinalizationGroup(_fg_)
    1. Return HostEnqueueJob("whatever", CleanupFinalizationGroup, << _fg_ >>).
```
If I can "implement" a one-off callback using `HostEnqueueJob`, then a one-off callback doesn't seem to be more flexible. (If it's deemed to have expositional benefit, we'd be free to retain it but define it as above. Of course, then it wouldn't be a "Host" operation, so name-change.)

In terms of *flexibility*, the three approaches look about the same to me.

But in *other* ways, `HostEnqueueJob` seems to me to be better: less to specify, less to understand, and less for downstream specs to have to deal with.

So I'm probably missing something. But what?

Reply:
1. For Promises, there are several different algorithms that need to be run, scheduled in the same way. So there's still value in passing around an abstract algorithm, while having these callbacks per type of scheduling.

2. Yes, there are other ways of attaching invariants, and we could do it with HostEnqueueJob. We could have a string parameter, and specify the invariants to be related to what the string is. I see this as an extremely minor editorial difference compared to the patch under discussion.Tor me, it feels marginally cleaner to attach invariants to a whole algorithm rather than to the particular string. But we could go the other way; it just isn't a big deal. You probably aren't missing much.

I don't really understand why HostEnqueueJob would be less to specify or understand. There would probably be more lines of spec text, since we'd have to break things up by the queue name on both the JS and host environment side anyway. It would be sort of like, in a computer program, merging a few functions together with a switch statement at the top, so that we have fewer functions that you have to think about.

Reply:
Re 1: I think I agree.

>     2. I see this as an extremely minor editorial difference

I agree that it's editorial. But given that it's an interface that downstream specs will probably reference, I don't think it's "extremely minor".
 
> compared to the patch under discussion.To me, it feels marginally cleaner to attach invariants to a whole algorithm rather than to the particular string. But we could go the other way; it just isn't a big deal.

 
> 
> I don't really understand why HostEnqueueJob would be less to specify or understand.

Because it's one operation rather than many.

> There would probably be more lines of spec text,

I doubt it. Each new HostEnqueueFooJob operation would have some boilerplate overhead.

> since we'd have to break things up by the queue name on both the JS and host environment side anyway.

Not quite sure what you mean.

> It would be sort of like, in a computer program, merging a few functions together with a switch statement at the top, so that we have fewer functions that you have to think about.

Sort of.